sympy/functions/special/hyperbolastic_functions.py

```diff
@@ -0,0 +1,161 @@
+    distance from a symmetric sigmoidal curve.
+
+    Examples
+    ========
+
+    >>> from sympy import diff, symbols, h1
+    >>> a ,b, c, x, y, z = symbols('M delta theta x0 P0 x')
+
+    >>> h1(a, b, c, x, y, z)
+    M*P0/(P0 + (M - P0)*exp(-delta*x + delta*x0 - theta*asinh(x) - theta*asinh(x0)))
+
+    >>> h1(a, b, 0, 1, 8, z)
+    M*exp(delta*x)/((M - 8)*exp(delta)/8 + exp(delta*x))
+
+    >>> diff(h1(a, b, 0, x, y, z), z).simplify()
+    M*P0*delta*(M - P0)*exp(delta*(x + x0))/(P0*exp(delta*x) + (M - P0)*exp(delta*x0))**2
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Hyperbolastic_functions
+    .. [2] https://tbiomed.biomedcentral.com/articles/10.1186/1742-4682-2-14
+
+    """
+
+    def fdiff(self, argindex=1):
+        if argindex == 6:
+            M, delta, theta, x0, P0 , x= self.args
+            p_x = h1(M, delta, theta, x0, P0 , x)
+            print(p_x)
+            return (p_x/M*(M - p_x)*(delta + theta/(sqrt(1 + x**2)))).simplify()
+        else:
+            raise ArgumentIndexError(self, argindex)
+
+    @classmethod
+    def eval(cls ,M, delta, theta, x0, P0 , x):
+        alpha = (M - P0)/P0*exp(delta*x0 - theta*asinh(x0))
+        return M/(1 + alpha*exp(-delta*x - theta*asinh(x))).simplify()
+
+class h2(Function):
+    r"""
+    The hyperbolastic rate equation of type II, denoted H2, is given by:
+
+    .. math::
+       \frac{\mathrm{d}P(x)}{\mathrm{d}x} = \frac{\delta\alpha\gamma P(x)^2 x^{\gamma-1}\tanh\left(\frac{M-P(x)}{\alpha P(x)}\right)}{M}
+
+    with initial condition P(t0) = P0 and gamma > 0, where tanh stands for hyperbolic
+    tangent function, M is the carrying capacity, and beta and gamma are parameters.
+    As in the H1 model, parameter beta has to be positive for increasing growth
+    curves with an asymptote at M and is negative only for decay profiles.
+    We refer to the growth rate model above as the hyperbolastic differential
+    equation of type II.
+
+
+    Examples
+    ========
+
+    >>> from sympy import h2, symbols
+    >>> a ,b, c, x, y, z = symbols('M delta gamma x0 P0 x')
+
+    >>> h2(a, b, c, x, y, z)
+    M*P0/(P0 + (M - P0)*asinh(exp(-delta*gamma*x))*asinh(exp(-delta*gamma*x0)))
+
+    >>> h2(a, 0, c, x, y, z)
+    M*P0/(P0 + (M - P0)*log(1 + sqrt(2))**2)
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Hyperbolastic_functions
+    .. [2] https://tbiomed.biomedcentral.com/articles/10.1186/1742-4682-2-14
+
+    """
+
+    def fdiff(self, argindex=1):
+        if argindex == 6:
+            M, delta, gama, x0, P0 , x= self.args
+            p_x = h2(M, delta, gama, x0, P0 , x)
+            alpha = (M - P0)/P0*asinh(exp(-delta*x0*gama))
+            return (p_x**2*x**(gama-1)*gama*delta*alpha)/M*tanh((M - p_x) \
+                                                    /alpha*p_x).simplify()
+        else:
+            raise ArgumentIndexError(self, argindex)
```

Comment:
Would `NotImplementedError` make more sense here, for when the index is valid but you don't know how to differentiate it?

Reply:
@eric-wieser yes ig NotImplementedError would make more sense . I've made the suggested changes please review it.